pkg/kubectl/cmd/auth.go

```diff
@@ -0,0 +1,451 @@
+
+For this command to work, "current-context" must be set in the kubeconfig, and that context must have an "oidc" section with a valid "refresh-token" inside of it. 
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunAuthRefresh(out, cmd, f, configAccess, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+	return cmd
+}
+
+// NewCmdAuthLogin creates a command object for the "auth login" command, which authenticates a user via the browser and stores the obtained token in the kubeconfig.
+func NewCmdAuthLogin(out io.Writer, f *cmdutil.Factory, configAccess cmdconfig.ConfigAccess) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "login",
+		Short: "login obtains authentication credentials to the cluster via the browser",
+		Long: `login will spawn the system's default browser, and navigate it to a page which allow the user to authenticate and obtain user credentials for interacting with the cluster.
+
+The authentication token will be stored in the current-context's user's "token" field. If a refresh-token is obtained, it will be stored in the user's oidc's section "refresh-token" field.
+ `,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunAuthLogin(out, cmd, f, configAccess, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+	return cmd
+}
+
+func RunAuthRefresh(out io.Writer, cmd *cobra.Command, f *cmdutil.Factory, configAccess cmdconfig.ConfigAccess, args []string) error {
+	cfg, err := configAccess.GetStartingConfig()
+	if err != nil {
+		return err
+	}
+
+	// Make sure there's a user.
+	user, err := GetCurrentAuthInfo(cfg)
+	if err != nil {
+		return err
+	}
+
+	// Ensure that the AuthInfo has an OIDC
+	if user.OIDCInfo == nil {
+		return fmt.Errorf("The current-context's user has no oidc information")
+	}
+
+	// Ensure that the OIDC has a refresh token
+	if user.OIDCInfo.RefreshToken == "" {
+		return errors.New("The current-context's user has no oidc refresh token")
+	}
+
+	// Ensure that a cluster is set.
+	cluster, err := GetCurrentCluster(cfg)
+	if err != nil {
+		return err
+	}
+
+	authClient, err := NewOIDCAuthClient(cluster)
+	if err != nil {
+		return err
+	}
+
+	// Do the exchange.
+	idToken, err := authClient.ExchangeRefreshToken(user.OIDCInfo.RefreshToken)
+	if err != nil {
+		return err
+	}
+	if idToken == "" {
+		return errors.New("No id_token returned from API Server")
+	}
+
+	// Store the obtained token in the config.
+	user.Token = idToken
+	err = cmdconfig.ModifyConfig(configAccess, *cfg, false)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func RunAuthLogin(out io.Writer, cmd *cobra.Command, f *cmdutil.Factory, configAccess cmdconfig.ConfigAccess, args []string) error {
+	cfg, err := configAccess.GetStartingConfig()
+	if err != nil {
+		return err
+	}
+
+	cluster, err := GetCurrentCluster(cfg)
+	if err != nil {
+		return err
+	}
+
+	user, err := GetCurrentAuthInfo(cfg)
+	if err != nil {
+		return err
+	}
+
+	authClient, err := NewOIDCAuthClient(cluster)
```

Comment:
`login` is a very general command... it shouldn't assume oidc on the server. This is where I'd expect to discover the auth method from the server, and branch into a basic challenge flow, negotiate challenge flow, or browser-based token flow. It's fine to not build all of those now, but we should have enough of the hooks in place so that an old kubectl knows whether or not it supports the auth method being used by the server.